xamarin/Xamarin.Forms
# #7956 [Shell] Unable to focus and unfocus search handler for Android

### Description of Change ###

This pull request modifies the ShellSearchView class in order to actually trigger the SearchHandler OnFocused() and OnUnfocused() overridable methods as well as the Focused and Unfocused EventHandlers (disregarding if they were declared on XAML or in the CodeBehind) whenever the TextBox within the SearchHandler changes its focus state. 
It also adds the Github Issues tests 7956 and 7956_1, in order to properly test such triggering of events.

The original commit that introduced the issue was e2318822 . After having analised the code, my guess is that such commit intended to implement most of the features already existing within the VisualElement class by replicating parts of its code within the SearchHandler class (rather than by extending it, as many other controls do either directly or indirectly).

The mentioned commit ammended existing test cases based on the SearchHandlerPage and StorePages in order to test properties like BackgroundColor, CancelButtonColor, TextColor, PlaceholderColor, HorizontalTextAlignment, Keyboard, FontFamily, etc. but in none of them it tests that the Focused or Unfocused event handlers trigger. 


### Issues Resolved ### 
- fixes #7956 

### API Changes ###

Added:
 - protected virtual void ShellSearchView.OnSearchHandlerFocusChange(object sender, FocusChangeEventArgs e);

Changed:
 - protected virtual void ShellSearchView.LoadView(SearchHandler searchHandler);
 
 Removed:
 None.
 
 
### Platforms Affected ### 

- Android

### Behavioral/Visual Changes ###

In case a SearchHandler (or an inheriting class) is declared within XAML providing Focused and/or Unfocused EventHandler functions, upgrading to this version will result in such events to actually trigger.
 

### Before/After Screenshots ### 

Before: The SearchHandler.cs Focused and/or Unfocused overridable functions nor the  Focused and/or Unfocused EventHandlers (no matter if declared in XAML or CodeBehind) wouldn't trigger. Actions within the event handlers wouldn't take place (in this test, the labels contents wouldn't be shown).
![image](https://user-images.githubusercontent.com/74742515/123721077-ca3a4b00-d885-11eb-93cb-b4bdb90d2232.png)

After:

The SearchHandler.cs Focused and/or Unfocused overridable functions **and** the  Focused and/or Unfocused EventHandlers (no matter if declared in XAML or CodeBehind) **trigger**. Actions within the event handlers **take place** (in this test, the labels contents **are** shown)

Issue7956.xaml - Focus
![image](https://user-images.githubusercontent.com/74742515/123798986-07cfc000-d8e8-11eb-8963-4e349978296b.png)
![image](https://user-images.githubusercontent.com/74742515/123799057-1ae29000-d8e8-11eb-8576-c54add6bb7c4.png)
![image](https://user-images.githubusercontent.com/74742515/123799209-436a8a00-d8e8-11eb-860e-c54bd20d455b.png)

Issue7956.xaml - Unfocus
![image](https://user-images.githubusercontent.com/74742515/123799262-53826980-d8e8-11eb-908b-81528d45dc86.png)
![image](https://user-images.githubusercontent.com/74742515/123799305-609f5880-d8e8-11eb-9968-430510b7db51.png)
![image](https://user-images.githubusercontent.com/74742515/123799393-790f7300-d8e8-11eb-81b2-593ca354e285.png)

Issue7956_1.xaml - Focus
![image](https://user-images.githubusercontent.com/74742515/123799659-ba078780-d8e8-11eb-8be6-c519e1176b26.png)
![image](https://user-images.githubusercontent.com/74742515/123799700-c4c21c80-d8e8-11eb-8168-06e9b1b57575.png)
![image](https://user-images.githubusercontent.com/74742515/123801587-b70d9680-d8ea-11eb-81b8-c89a5ed9978e.png)
![image](https://user-images.githubusercontent.com/74742515/123801725-d5739200-d8ea-11eb-9f81-f606ccb78b6b.png)

Issue7956_1.xaml - Unfocus
![image](https://user-images.githubusercontent.com/74742515/123801821-ec19e900-d8ea-11eb-808c-591bf2dbfcf8.png)
![image](https://user-images.githubusercontent.com/74742515/123801884-f805ab00-d8ea-11eb-948b-36899c33ae11.png)
![image](https://user-images.githubusercontent.com/74742515/123801923-03f16d00-d8eb-11eb-855f-6f7ff229f29a.png)
![image](https://user-images.githubusercontent.com/74742515/123802003-18356a00-d8eb-11eb-9dda-baa07d7417e7.png)


### Testing Procedure ###

1. Launch any of the Xamarin.Forms.ControlGallery.[Platform]
2. On the main App screen, hit the 'GO TO TEST CASES'
3. Search for the Test case 7956 and open it.
4. A Shell view appears having a SearchHandler control on top, also a descriptive text explaining the test itself.
5. Click (or poke) on the SearchHandler in order to cause the Focus event to trigger. A text should appear saying: Focus EventHandler triggered!
6. Click (or poke) on any other area of the screen but the SearchHandler in order to cause the Unfocus event to trigger. A text should appear saying: Unfocus EventHandler triggered!

If steps 5 or 6 produce the mentioned output, the test would have passed. Otherwise, it would have failed. 


### PR Checklist ###
